benchmarks: add lucbv's deep_copy benchmark

cwpearson opened the pull request:
**Please see https://github.com/kokkos/kokkos/pull/7780#issuecomment-2672134470 for revised approach**

<hr>

Add a deep_copy benchmark to facilitate Trilinos performance investigations. Mostly authored by @lucbv.

<details>
<summary>run.sh</summary>

```bash
#! /bin/bash

KOKKOS_SRC="${PWD}"
KOKKOS_BUILD="${PWD}/build-7780"

set -eou pipefail

if true; then
rm -rf "$KOKKOS_BUILD"
cmake -S "$KOKKOS_SRC" -B "$KOKKOS_BUILD" \
  -DKokkos_ENABLE_SERIAL=ON \
  -DKokkos_ENABLE_BENCHMARKS=ON
fi
cmake --build "$KOKKOS_BUILD" --parallel $(nproc)

"$KOKKOS_BUILD"/benchmarks/deep_copy/Kokkos_deep_copy
```

</details>

<details>
<summary>Example OpenMP + CUDA output</summary>

```
-------------------------------------------------------------
Kokkos deep_copy Benchmark
-------------------------------------------------------------
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false

Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- View Size:    100000000
- Per View:           800.00 MB
Benchmark kernels will be performed for 20 iterations.
Note: the strided case copies half the view.
-------------------------------------------------------------
Start benchmarking for contiguous layout...
-------------------------------------------------------------
Scalar to Device              2.99 ms         255180.94 MiB/s
Scalar to Host                7.15 ms         106758.24 MiB/s
Host to Device               78.98 ms           9659.37 MiB/s
Device to Host               42.62 ms          17901.36 MiB/s
Host to Host                 13.88 ms          54950.50 MiB/s
Device to Device              6.15 ms         124018.13 MiB/s
-------------------------------------------------------------
Start benchmarking for strided layout...
-------------------------------------------------------------
Scalar to Device              3.46 ms         110112.36 MiB/s
Scalar to Host               12.85 ms          29692.99 MiB/s
Host to Host                 19.68 ms          19379.85 MiB/s
Device to Device              6.20 ms          61532.33 MiB/s
-------------------------------------------------------------
```
</details>

<details>
<summary>Example Serial output</summary>

```
-------------------------------------------------------------
Kokkos deep_copy Benchmark
-------------------------------------------------------------
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- View Size:    100000000
- Per View:           800.00 MB
Benchmark kernels will be performed for 20 iterations.
Note: the strided case copies half the view.
-------------------------------------------------------------
Start benchmarking for contiguous layout...
-------------------------------------------------------------
Scalar to Device             54.82 ms          13917.51 MiB/s
Scalar to Host               54.40 ms          14023.79 MiB/s
Host to Device                0.00 ms     8477105034.72 MiB/s
Device to Host                0.00 ms     8572353405.90 MiB/s
Host to Host                 67.94 ms          11230.19 MiB/s
Device to Device             67.91 ms          11234.60 MiB/s
-------------------------------------------------------------
Start benchmarking for strided layout...
-------------------------------------------------------------
Scalar to Device             57.61 ms           6621.73 MiB/s
Scalar to Host               57.40 ms           6645.39 MiB/s
Host to Host                 78.29 ms           4872.60 MiB/s
Device to Device             77.95 ms           4893.84 MiB/s
-------------------------------------------------------------
```
</details>

lucbv commented on benchmarks/deep_copy/Makefile:
If adding the Makefile is not really required then I would suggest removing it since it is being deprecated soon?

lucbv reviewed (approved):
I am in favor, I think this adds useful coverage especially to ensure that performance on apu vs cpu+gpu is what we expect it to be.
It also informs us on optimization like using memset vs parallel_for

masterleinad commented:
Where is this motivated from?

lucbv commented:
@masterleinad please read the comment I left just above your question

dalg24 commented on benchmarks/deep_copy/deep_copy.cpp:
Consider using Google Benchmark.
We have been trying to transition to it instead of rolling our own.

dalg24 commented on benchmarks/deep_copy/deep_copy.cpp:
```diff
@@ -0,0 +1,154 @@
+#include "Kokkos_Core.hpp"
+#include <cstdio>
+
+#define COPY_ARRAY_SIZE 100000000
+#define COPY_NTIMES 20
+
+#define HLINE "-------------------------------------------------------------\n"
```
Don't use macro for these.

cwpearson commented:
I reworked this to re-use some machinery in the existing perf_tests:
* modified the rank-1 deep_copy tests to accept a memory space, so we can do host/device copy benchmarks
* added a strided version of view_fill
* added a strided version of deep copy

Questions:
1. With the current implementation, the google benchmark names can be kind of long:
```
ViewDeepCopy_Rank1<Kokkos::LayoutLeft, Kokkos::LayoutLeft, Kokkos::DefaultExecutionSpace::memory_space, Kokkos::DefaultHostExecutionSpace::memory_space>/N:10/manual_time
```
Maybe I could alias that down to a shorter name where the benchmark is defined?
2. The host/default benchmarks will happen even if host and default are the same, and will end up being redundant to the default/default benchmarks. Should I find a way to avoid this?

crtrott commented:
Retest this please.
